greenplumpython/func.py

```diff
@@ -170,292 +158,325 @@ def serialize(self) -> str:
+# It is not a Callable by design to prevent misuse.
+class _AbstractFunction:
+    def __init__(
+        self,
+        wrapped_func: Optional[Callable[..., Any]],
+        name: Optional[str],
+        schema: Optional[str],
+    ) -> None:
+        NAMEDATALEN = 64  # See definition in PostgreSQL
+        if not (
+            (name is not None and len(name) < NAMEDATALEN)
+            or (wrapped_func is not None and len(wrapped_func.__name__) < NAMEDATALEN)
+        ):
+            raise Exception(f"Function name should be shorter than {NAMEDATALEN} bytes.")
+        qualified_name = (
+            (name if schema is None else f"{schema}.{name}")
+            if name is not None
+            else f"pg_temp.{wrapped_func.__name__}"
+            if wrapped_func is not None
+            else None
+        )
+        assert qualified_name is not None
+        assert (
+            qualified_name not in _global_scope
+        ), f'Function named "{qualified_name}" has been defined before.'
+        self._qualified_name = qualified_name
+        _global_scope[qualified_name] = self
 
-    Example:
-            .. code-block::  Python
+    @property
+    def qualified_name(self) -> str:
+        return self._qualified_name
+
+    def _try_resolve_db(self, args: Tuple[Any, ...], db: Optional[Database]) -> Optional[Database]:
+        if db is None:
+            for arg in args:
+                if isinstance(arg, Expr) and arg.db is not None:
+                    db = arg.db
+                    break
+        return db
 
-                generate_series = gp.function("generate_series", db)
 
-    """
+_global_scope: Dict[str, _AbstractFunction] = {}
 
-    def make_function_call(
-        *args: Expr, as_name: Optional[str] = None, db: Optional[Database] = None
-    ) -> FunctionExpr:
-        return FunctionExpr(name, args, as_name=as_name, db=db)
 
-    return make_function_call
+class NormalFunction(_AbstractFunction):
+    def __init__(
+        self,
+        wrapped_func: Optional[Callable[..., Any]] = None,
+        name: Optional[str] = None,
+        schema: Optional[str] = None,
+        returning_composite: bool = False,
+        language_handler: str = "plpython3u",
+    ) -> None:
+        super().__init__(wrapped_func, name, schema)
+        self._created_in_dbs: Optional[Set[Database]] = set() if wrapped_func is not None else None
+        self._wrapped_func = wrapped_func
+        self._language_handler = language_handler
+        self._returning_composite = returning_composite
+
+    def unwrap(self) -> Callable[..., Any]:
+        """
+        Get the wrapped Python function in the database function.
+        """
+        assert self._wrapped_func is not None, "No Python function is wrapped inside."
+        return self._wrapped_func
+
+    def _create_in_db(self, db: Database) -> None:
+        assert self._created_in_dbs is not None
+        assert self._wrapped_func is not None
+        if db not in self._created_in_dbs:
```

Comment:
Do we really want to handle multiple database cases (in the future)? from the implementation of `_try_resolve_db()`, it is not handled well.

Reply:
`_try_resolve_db()` always get the first `db` to use, what if the others are not the same with the first one?